File: c/slvSys.c
/* -*- Mode:C; Coding:us-ascii-unix; fill-column:158 -*- */
/*******************************************************************************************************************************************************.H.S.**/
/**
 @file      slvSysC.c
 @author    Mitch Richling http://www.mitchr.me/
 @brief     Demonstrate LAPACK's sgesv functions.@EOL
 @keywords  lapacke claback cblas sgesv numerical linear algebra
 @std       C99
 @see       https://github.com/richmit/ex-blas-lapack/
 @copyright 
  @parblock
  Copyright (c) 1998,2025, Mitchell Jay Richling <http://www.mitchr.me/> All rights reserved.
  
  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
  
  1. Redistributions of source code must retain the above copyright notice, this list of conditions, and the following disclaimer.
  
  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions, and the following disclaimer in the documentation
     and/or other materials provided with the distribution.
  
  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software
     without specific prior written permission.
  
  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
  OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
  DAMAGE.
  @endparblock
 @todo      @EOL@EOL
 @warning   @EOL@EOL
 @bug       @EOL@EOL
 @filedetails   

  Demonstrate LAPACK's Xgesv function as exposed in the LAPACKE C interface.  Note that you can still use the claback calls.
*/
/*******************************************************************************************************************************************************.H.E.**/

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
#include <stdio.h>                                                       /* I/O lib                 C89      */
#include <stdlib.h>                                                      /* Standard Lib            C89      */
#include <cblas.h>                                                       /* C BLAS                  BLAS     */
#include <lapacke.h>                                                     /* LAPACK C Interface      LAPACKE  */

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
#include "blaio.h"                                                       /* Basic Linear Algebra I/O         */

/*------------------------------------------------------------------------------------------------------------------------------------------------------------*/
int main(int argc, char **argv) {
  lapack_int n       = 4;
  lapack_int nrhs    = 1;
  lapack_int lda     = n;
  lapack_int ldb     = n;
  double a[]         = {   1.0,  2.0,  3.0,  4.0,  /* n*n */
                           6.0,  7.0,  9.0,  9.0,
                          11.0, 12.0, 19.0, 14.0,
                          16.0, 17.0, 18.0, 12.0,
                        };
  double b[]         = {1.0, 3.0, 5.0, 6.0};       /* n*nrhs */
  lapack_int ipiv[]  = {  0,   0,   0,   0};       /* n */
  lapack_int inf;
  int i;

  printMatrix(CblasColMajor, 4, 4, a, 15, 10, NULL, NULL, NULL, NULL, NULL, "a=");
  printMatrix(CblasColMajor, 1, 4, b, 15, 10, NULL, NULL, NULL, NULL, NULL, "b=");

  /* The LAPACKE version of dgesv */
  inf = LAPACKE_dgesv(LAPACK_COL_MAJOR, n, nrhs, a, lda, ipiv, b, ldb);
  /* We can use the lapacke header with old clapack code too.  Example: */
  // dgesv_(&n, &nrhs, a, &lda, ipiv, b, &ldb, &inf);

  if(inf == 0) {
    printf("Successful Solution\n");
  } else if(inf < 0) {
    printf("Illegal value at: %d\n", -(int)inf);
    exit(1);
  } else if(inf > 0) {
    printf("Matrix was singular\n");
    exit(1);
  } else {
    printf("Unknown Result (Can't happen!)\n");
    exit(1);
  } /* end if/else */

  printMatrix(CblasColMajor, 4, 4, a, 15, 10, NULL, NULL, NULL, NULL, NULL, "a=");
  printMatrix(CblasColMajor, 1, 4, b, 15, 10, NULL, NULL, NULL, NULL, NULL, "b=");

  printf("PIV=");
  for(i=0;i<4;i++)
    printf("%4d ", (int)(ipiv[i]));
  printf("\n");

  return 0;

} /* end func main */

File: f77/slvSys.f
C -*- Mode:Fortran; Coding:us-ascii-unix; fill-column:72 -*-
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC.H.S.CC
C @file      slvSys.f
C @author    Mitch Richling http://www.mitchr.me/
C @brief     Solve a system of equations with DGESV. @EOL
C @keywords  blas lapack numerical linear algebra matrix vector netlib
C @std       F77 MIL-STD-1753
C @see       https://github.com/richmit/ex-blas-lapack/
C @copyright
C  @parblock
C  Copyright (c) 1996,2025, Mitchell Jay Richling All rights reserved.
C
C  Redistribution and use in source and binary forms, with or without
C  modification, are permitted provided that the following conditions
C  are met:
C
C  1. Redistributions of source code must retain the above copyright
C     notice, this list of conditions, and the following disclaimer.
C
C  2. Redistributions in binary form must reproduce the above
C     copyright notice, this list of conditions, and the following
C     disclaimer in the documentation and/or other materials provided
C     with the distribution.
C
C  3. Neither the name of the copyright holder nor the names of its
C     contributors may be used to endorse or promote products derived
C     from this software without specific prior written permission.
C
C  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
C  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
C  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
C  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
C  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
C  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
C  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
C  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
C  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
C  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
C  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
C  OF THE POSSIBILITY OF SUCH DAMAGE.
C  @endparblock
C @filedetails
C
C  Demonstrates the DGESV routine from LAPACK to solve a system of
C  equations.
C
CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC.H.E.CC

C-----------------------------------------------------------------------
      program slvSysF

      implicit none

      double precision a(4,4), b(4)
      integer pivs(4), inf

C     Externals from the blaio library
      external dgeprt
C     Externals from the LAPACK library
      external dgesv

C     Initialize the matrix a and the vector b
      data a/ 1, 2, 3, 4,
     *        6, 7, 9, 9,
     *       11,12,19,14,
     *       16,17,18,12/
      data b/ 1, 3, 5, 6/

C     Print out the matrix and vector we start with
      call dgeprt(4, 4, a, 'a=')
      call dgeprt(1, 4, b, 'b=')

C     Compute the solution
      call dgesv(4, 1, a, 4, pivs, b, 4, inf)

C     Figure out if dgesv found a solution or not
      if (inf .eq. 0) then
         write (*,*) 'Successful solution'
      else if (inf .lt. 0) then
         write (*,*) 'Illegal value at: ', -inf
         stop
      else if (inf .gt. 0) then
         write (*,*) 'matrix was singular'
         stop
      else
         write (*,*) 'Illegal return code'
         stop
      end if

C     Print out the answer
      call dgeprt(4, 4, a, 'a=')
      call dgeprt(1, 4, b, 'b=')
      write(*,*) 'pivs=', pivs

      end
